### PR TITLE
Testing generated projects with the various options

### PR DESCRIPTION
**Pull request by esciara**
This includes:

* cli options (bdd tests, formating): 
  * cli-cleo (default)
  * cli-click
  * cli-argparse
  * cli-none
* test library options:
  * pytest (default)
  * unittest
* hosting options:
  * gitlab (default)
  * github

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,32 +1,54 @@
...
-COOKIECUTTER_OUTPUT_DIR=$(BUILD_DIR)/a_generated_python_project
+COOKIECUTTER_OUTPUT_DIR=$(BUILD_DIR)/a-generated-python-project
 
 help:
 	@echo "help"
 
 setup-cicd:
 	pip install --upgrade pip
-	pip install cookiecutter
+	pip install -r requirements.txt
```
Comment (slequeux) [inline Makefile]: Why not use poetry for `requirements.txt` and `test_requirements.txt` ?

File: hooks/pre_gen_project.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+MODULE_REGEX = r'^[_a-zA-Z][_a-zA-Z0-9]+$'
+
+module_name = '{{ cookiecutter.project_package_name}}'
+
+if not re.match(MODULE_REGEX, module_name):
+    print('ERROR: The project package name (%s) is not a valid Python module name. Please do not use a - and use _ instead' % module_name)
+
+    #Exit to cancel project
+    sys.exit(1)
```
Comment (slequeux) [inline hooks/pre_gen_project.py:13]: add a `if __name__ == "__main__":` ?

File: tests/test_no_github_in_gitlab_generated_project.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        "github.com/thejohnfreeman/",
+        "github.com/python/",
+        "github.com/psf/",
+        "github.com/audreyr/",
+        "github.com/opinionated-digital-center/python-library-project-generator",
+    ]
+
+    for file in files:
+        if str(file.relative_to(project_path)) in files_to_exclude or not file.is_file():
+            continue
```
Comment (slequeux) [inline tests/test_no_github_in_gitlab_generated_project.py]: Use a list comprehension before the for to filter files ?

File: tests/test_no_github_in_gitlab_generated_project.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                        words = line.split(" ")
+                        for word in words:
+                            if not any([excluded_content in word for excluded_content in content_to_exclude]):
+                                failed = True
+                    else:
+                        if not any([excluded_content in line for excluded_content in content_to_exclude]):
+                            failed = True
+
+            if failed:
+                files_containing_github.append(str(file.relative_to(project_path)))
```
Comment (slequeux) [inline tests/test_no_github_in_gitlab_generated_project.py]: Extract a function that returns boolean to filter list ?

File: tests/test_no_github_in_gitlab_generated_project.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        with file.open() as f:
+            failed = False
+            lines = f.readlines()
+            for line in lines:
+                count = line.lower().count("github")
+                if count > 0:
+                    if count > 1:
+                        words = line.split(" ")
+                        for word in words:
+                            if not any([excluded_content in word for excluded_content in content_to_exclude]):
```
Comment (slequeux) [inline tests/test_no_github_in_gitlab_generated_project.py]: Just before applying the count, maybe remoe all occurences of `content to exclude` words so there will be no duplication of this line

File: {{cookiecutter.project_slug}}/Makefile [hunk 6]
```diff
@@ -49,10 +49,13 @@ help:
...
 	@echo "\tsetup-release-tools: - setup tools needed for releasing package"
-	@echo "\tsetup-cicd-test-stage-github: - setup tools needed for CI/CD test stage (GitHub)"
-	@echo "\tsetup-cicd-release-stage-github: - setup tools needed for CI/CD release stage (GitHub)"
+{%- if cookiecutter.hosting == "GitLab" %}
 	@echo "\tsetup-cicd-test-stage-gitlab: - setup tools needed for CI/CD test stage (GitLab)"
 	@echo "\tsetup-cicd-release-stage-gitlab: - setup tools needed for CI/CD release stage (GitLab)"
+{%- else %}
+	@echo "\tsetup-cicd-test-stage-github: - setup tools needed for CI/CD test stage (GitHub)"
+	@echo "\tsetup-cicd-release-stage-github: - setup tools needed for CI/CD release stage (GitHub)"
+{%- endif %}
```
Comment (slequeux) [inline {{cookiecutter.project_slug}}/Makefile]: `gitlab` and `github` qualifiers are not necessayr in make target name

File: {{cookiecutter.project_slug}}/Makefile [hunk 7]
```diff
@@ -130,11 +133,14 @@ setup-release-tools-common:
...
 	npm install -g @semantic-release/git@"^9.0.0"
 
+{% if cookiecutter.hosting == "GitLab" -%}
 setup-release-tools-gitlab: setup-release-tools-common
 	npm install -g @semantic-release/gitlab@"^6.0.3"
 
+{%- else %}
 setup-release-tools-github: setup-release-tools-common
 	npm install -g @semantic-release/github@"^7.0.5"
+{%- endif %}
```
Comment (slequeux) [inline {{cookiecutter.project_slug}}/Makefile]: `gitlab` and `github` qualifiers are not necessayr in make target name

File: {{cookiecutter.project_slug}}/Makefile [hunk 8]
```diff
@@ -175,11 +181,14 @@ setup-cicd-poetry:
 	. $(HOME)/.poetry/env && poetry config virtualenvs.create false
 	echo "WARNING: you still need to source $$HOME/.poetry/env to access poetry's executable"
 
+{% if cookiecutter.hosting == "GitLab" -%}
 setup-cicd-test-stage-gitlab: setup-cicd-poetry
 
 setup-cicd-release-stage-gitlab: setup-cicd-python3 setup-cicd-poetry setup-release-tools-gitlab
+{%- else %}
```
Comment (slequeux) [inline {{cookiecutter.project_slug}}/Makefile:176]: `setup-cicd-test-stage-github` not implemented

File: tests/test_no_github_in_gitlab_generated_project.py (hunk 3 above)
Comment (esciara) [reply tests/test_no_github_in_gitlab_generated_project.py]: Done

File: tests/test_no_github_in_gitlab_generated_project.py (hunk 4 above)
Comment (esciara) [reply tests/test_no_github_in_gitlab_generated_project.py]: No need in the rewritten code... have a look.

File: {{cookiecutter.project_slug}}/Makefile (hunk 8 above)
Comment (esciara) [reply {{cookiecutter.project_slug}}/Makefile:176]: No need... changed the code.

File: {{cookiecutter.project_slug}}/Makefile (hunk 7 above)
Comment (esciara) [reply {{cookiecutter.project_slug}}/Makefile]: removed.

File: {{cookiecutter.project_slug}}/Makefile (hunk 6 above)
Comment (esciara) [reply {{cookiecutter.project_slug}}/Makefile]: Done

File: tests/test_no_github_in_gitlab_generated_project.py (hunk 5 above)
Comment (esciara) [reply tests/test_no_github_in_gitlab_generated_project.py]: Done

File: hooks/pre_gen_project.py (hunk 2 above)
Comment (esciara) [reply hooks/pre_gen_project.py:13]: I checked: no need for it at all.

File: Makefile (hunk 1 above)
Comment (esciara) [reply Makefile]: Done

File: Makefile [hunk 9]
```diff
@@ -1,32 +1,141 @@
...
+	$(MAKE) -C $(COOKIECUTTER_OUTPUT_DIR) install-test-dependencies
+	$(MAKE) -C $(COOKIECUTTER_OUTPUT_DIR) test
+
+bdd-generated-project: refresh-$(OPTION)
+	$(MAKE) -C $(COOKIECUTTER_OUTPUT_DIR) install-bdd-dependencies
+	$(MAKE) -C $(COOKIECUTTER_OUTPUT_DIR) bdd
+
+format-generated-project: refresh-$(OPTION)
+	$(MAKE) -C $(COOKIECUTTER_OUTPUT_DIR) install-format-dependencies
+	# seed-isort-config needs a `git` repository to function
```
Comment (antoinemichaud) [inline Makefile]: to function -> to work (je crois)

File: tests/test_no_github_in_gitlab_generated_project.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
+from cookiecutter.main import cookiecutter
+from hamcrest import assert_that, empty, is_, not_
+
+FILES_TO_EXCLUDE = [
```
Comment (antoinemichaud) [inline tests/test_no_github_in_gitlab_generated_project.py]: -> IGNORED_FILES

File: tests/test_no_github_in_gitlab_generated_project.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    assert_that(paths, not_(empty()))
+
+    paths = [
+        path for path in paths
+        if str(path.relative_to(project_path)) not in FILES_TO_EXCLUDE and
+           path.is_file()
+    ]
+    files_containing_github = []
+
+    for file in paths:
```
Comment (antoinemichaud) [inline tests/test_no_github_in_gitlab_generated_project.py:52]: Un petit extract function de ces deux boucles imbriquées ?

File: tests/test_no_github_in_gitlab_generated_project.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        with file.open() as f:
+            file_content = f.read()
+
+            for to_exclude in CONTENTS_TO_EXCLUDE:
+                file_content = file_content.replace(to_exclude, "")
+
+            if "github" in file_content.lower():
+                files_containing_github.append(str(file.relative_to(project_path)))
```
Comment (antoinemichaud) [inline tests/test_no_github_in_gitlab_generated_project.py]: Un petit extract de fonction ?

File: Makefile (hunk 9 above)
Comment (esciara) [reply Makefile]: done

File: tests/test_no_github_in_gitlab_generated_project.py (hunk 12 above)
Comment (esciara) [reply tests/test_no_github_in_gitlab_generated_project.py]: done

File: tests/test_no_github_in_gitlab_generated_project.py (hunk 10 above)
Comment (esciara) [reply tests/test_no_github_in_gitlab_generated_project.py]: done

**Pull request merged**